src/CPUSolver.cpp

```diff
@@ -38,132 +38,52 @@ int CPUSolver::getNumThreads() {
-  FP_PRECISION* nu_sigma_f = material->getNuSigmaF();
-  FP_PRECISION* chi = material->getChi();
-  FP_PRECISION fission_source = 0.0;
-  FP_PRECISION scatter_source = 0.0;
-  FP_PRECISION total_source = 0.0;
-
-  /* Compute fission source for each group */
-  if (material->isFissionable()) {
-    for (int e=0; e < _num_groups; e++)
-      fission_source += _scalar_flux(fsr_id,e) * nu_sigma_f[e];
-
-    fission_source /= _k_eff;
-  }
-
-  for (int g=0; g < _num_groups; g++)
-    scatter_source += material->getSigmaSByGroupInline(g,energy_group-1)
-        * _scalar_flux(fsr_id,g);
-
-  /* Compute the total source */
-  total_source = (fission_source * chi[energy_group-1] + scatter_source) *
-      ONE_OVER_FOUR_PI;
-
-  return total_source;
-}
-
-
-/**
- * @brief Return a scalar flux array indexed by FSR IDs and energy groups.
- * @details This energy groups are the innermost index, while the FSR ID is
- *         the outermost index.
- * @return an array of flat source region scalar fluxes
+ * @brief Sets the number of shared memory OpenMP threads to use (>0).
+ * @param num_threads the number of threads
  */
-FP_PRECISION* CPUSolver::getFSRScalarFluxes() {
+void CPUSolver::setNumThreads(int num_threads) {
 
-  if (_scalar_flux == NULL)
-    log_printf(ERROR, "Unable to returns the Solver's FSR scalar flux array "
-               "since it has not yet been allocated in memory");
+  if (num_threads <= 0)
+    log_printf(ERROR, "Unable to set the number of threads to %d "
+               "since it is less than or equal to 0", num_threads);
 
-  return _scalar_flux;
+  /* Set the number of threads for OpenMP */
+  _num_threads = num_threads;
+  omp_set_num_threads(_num_threads);
 }
 
 
 /**
- * @brief Sets the number of shared memory OpenMP threads to use (>0).
- * @param num_threads the number of threads
+ * @brief Assign a fixed source for a flat source region and energy group.
+ * @details Fixed sources should be scaled to reflect the fact that OpenMOC 
+ *          normalizes the scalar flux such that the total energy- and 
+ *          volume-integrated production rate sums to 1.0.
+ * @param fsr_id the flat source region ID
+ * @param group the energy group
+ * @param source the volume-averaged source in this group
  */
-void CPUSolver::setNumThreads(int num_threads) {
+void CPUSolver::setFixedSourceByFSR(int fsr_id, int group, 
+                                    FP_PRECISION source) {
 
-  if (num_threads <= 0)
-    log_printf(ERROR, "Unable to set the number of threads for the Solver "
-               "to %d since it is less than or equal to 0", num_threads);
+  Solver::setFixedSourceByFSR(fsr_id, group, source);
 
-  _num_threads = num_threads;
+  /* Allocate the fixed sources array if not yet allocated */
+  if (_fixed_sources == NULL) {
+    int size = _num_FSRs * _num_groups;
+    _fixed_sources = new FP_PRECISION[size];
+    memset(_fixed_sources, 0.0, sizeof(FP_PRECISION) * size);
+  }
```

Comment:
This is a little simpler than my original implementation, as it keeps only a single fixed source array.  This is of course at the expense of memory, as I only allocated fixed source arrays for regions where it was nonzero.  All in all, I think I agree the simpler approach is the way to go, but we should keep in mind this is what is happening if running with a large problem spatially and in energy.